### PR TITLE
Add Github Actions CI 

### PR DESCRIPTION
**Pull request by HarrisonWAffel**
This PR adds CI for PR's, merges to `main`, and tag actions using github actions.

Some specifics 

- `.gitattributes` was added to enforce `lf` line endings on all files 
- Chart validations and tests have been disabled for windows runners 
  - Ran into an issue with Hull on Windows which prevented chart tests from running. [A PR was raised with a fix]( https://github.com/rancher/hull/pull/42)
    - I'm not sure it even makes sense to run the chart tests  / validations on each runner, as I couldn't find a case where the tests would behave differently on different OS's. Right now they are running on the ubuntu runner only.
  - Ran into various issues with the chart build scripts on Windows. The largest issue being the embedded `go-git` logic reporting file changes which did not exist and could not be reproduced when using `git status --porcelain` directly 

- `scripts/ci` is no longer being used in favor of multi-job github actions
- `scripts/publish` has been added which pushes images and manifests
   -  Merges into main will publish images with the format of `<APPLICATION_NAME>-<COMMIT>-<ARCH>`, and manifests with the format of `<APPLICATION_NAME>-<COMMIT>`. An example manifest can be found [here](https://hub.docker.com/layers/harrisonwaffel/gmsa-account-provider/b77ee84-amd64/images/sha256-319594c6b1faa16a914b54cbd93ed8cf7cc110a24960f6d393352bb98333b7e9?context=explore)
- two tag workflows have been added, one for prereleases and one for full releases. Both use the `marvinpinto/action-automatic-releases` action, which is reccomended by github in its [archived release action repo](https://github.com/actions/create-release)

I've tested the image pushing / manifest creation steps on my local fork with my local docker hub account. Forks can use the release step if needed by specifying the github action `REPO`  variable, as well as the `DOCKER_USERNAME` and `DOCKER_PASSWORD` github action secrets. In the event that `REPO` is not set, it will default to `rancher`. The tag workflows have not been tested yet. 

Additional resources: 
A list of installed programs present on all Windows runners can be found [here](https://github.com/actions/runner-images/blob/main/images/win/Windows2019-Readme.md)

### CODE REVIEW

File: scripts/ci [hunk 1]
```diff
@@ -3,11 +3,10 @@ set -e
 
 cd $(dirname $0)
 
-./build
+./validate-ci
```
Comment (aiyengar2) [inline scripts/ci:6]: this needs to be kept at the end. It's a validation that none of the other scripts that are executed produce artifacts after execution

File: scripts/package [hunk 2]
```diff
@@ -20,24 +20,14 @@ for i in "${IMAGES[@]}"; do
   IMAGE=${REPO}/"$i":${TAG}
   DOCKERFILE="package/$i/Dockerfile${TEMP_DOCKER_SUFFIX}"
 
-  FINAL_IMAGES=()
   if [ "$OS" = "windows" ]; then
       for v in "${NANOSERVER_VERSIONS[@]}"; do
         # append the nano-server version to the image tag
         IMAGE=${REPO}/"$i":${TAG}-${v}
+        echo "docker build --build-arg "ARCH=${ARCH}" --build-arg "OS=${OS}" --build-arg="NANOSERVER_VERSION=${v}" -f ${DOCKERFILE} -t ${IMAGE}  ."
```
Comment (aiyengar2) [inline scripts/package:27]: maybe wrap this in:

```bash
(
    set -x
    docker build --build-arg "ARCH=${ARCH}" --build-arg "OS=${OS}" --build-arg="NANOSERVER_VERSION=${v}" -f ${DOCKERFILE} -t ${IMAGE} .
)

File: scripts/validate [hunk 3]
```diff
@@ -1,21 +1,10 @@
...
-
-if ! command -v golangci-lint; then
-    echo Skipping validation: no golangci-lint available
-    exit
-fi
-
-echo Running validation
-
-echo Running: golangci-lint
-golangci-lint run
```
Comment (aiyengar2) [inline scripts/validate:18]: why did we remove this?

**Review by aiyengar2 — CHANGES_REQUESTED** [review]

File: scripts/validate (hunk 3 above)
Comment (HarrisonWAffel) [reply scripts/validate:18]: I removed it in favor of the golangci-lint github action. I can add this back in and simply not call the file, and run `go fmt ./...` directly from the action

File: scripts/package (hunk 2 above)
Comment (HarrisonWAffel) [reply scripts/package]: added

File: scripts/ci (hunk 1 above)
Comment (HarrisonWAffel) [reply scripts/ci]: reverted

File: scripts/validate (hunk 3 above)
Comment (HarrisonWAffel) [reply scripts/validate:18]: reverted

File: scripts/ci [hunk 4]
```diff
@@ -6,8 +6,7 @@ cd $(dirname $0)
 ./build
 ./test
 ./test-charts
-# todo; during validation, git is going to report file changes due to different line endings
 ./validate
 ./validate-charts
 ./validate-ci
-./package
+./package
```
Comment (aiyengar2) [inline scripts/ci:12]:
```suggestion
./package

```

File: scripts/package [hunk 5]
```diff
@@ -20,24 +20,17 @@ for i in "${IMAGES[@]}"; do
...
-  FINAL_IMAGES=()
   if [ "$OS" = "windows" ]; then
       for v in "${NANOSERVER_VERSIONS[@]}"; do
         # append the nano-server version to the image tag
         IMAGE=${REPO}/"$i":${TAG}-${v}
+        (
+            set -x
+            docker build --build-arg "ARCH=${ARCH}" --build-arg "OS=${OS}" --build-arg="NANOSERVER_VERSION=${v}" -f ${DOCKERFILE} -t ${IMAGE} .
+        )
         docker build --build-arg "ARCH=${ARCH}" --build-arg "OS=${OS}" --build-arg="NANOSERVER_VERSION=${v}" -f ${DOCKERFILE} -t ${IMAGE}  .
```
Comment (aiyengar2) [inline scripts/package:31]: No need to build twice.

```suggestion
        (
            set -x
            docker build --build-arg "ARCH=${ARCH}" --build-arg "OS=${OS}" --build-arg="NANOSERVER_VERSION=${v}" -f ${DOCKERFILE} -t ${IMAGE} .
        )
```

File: scripts/validate [hunk 6]
```diff
@@ -18,4 +18,4 @@ echo Running: golangci-lint
 golangci-lint run
 
 echo Running: go fmt
-test -z "$(go fmt ${PACKAGES} | tee /dev/stderr)"
+test -z "$(go fmt ${PACKAGES} | tee /dev/stderr)"
```
Comment (aiyengar2) [inline scripts/validate:21]:
```suggestion
test -z "$(go fmt ${PACKAGES} | tee /dev/stderr)"

```

**Review by aiyengar2 — CHANGES_REQUESTED** [review]
two nits, one change

File: scripts/package (hunk 5 above)
Comment (HarrisonWAffel) [reply scripts/package:31]: Ah right, one is supposed to be a log message

**Comment by HarrisonWAffel** [comment]
still a few minor changes being tested here

**Review by aiyengar2 — COMMENTED** [review]
The only thing that needs to be changed is the fact that we never run `./scripts/validate-ci` after these workflows. This is necessary to ensure that the changes committed into the repository match what we would produce in a full CI run.

Can you add this step to each of the workflows?

**Review by aiyengar2 — APPROVED** [review]

**Pull request merged**